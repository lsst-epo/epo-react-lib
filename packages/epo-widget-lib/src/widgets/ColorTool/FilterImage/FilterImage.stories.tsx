import { ComponentMeta, ComponentStoryObj } from "@storybook/react";
import { argTypes } from "@/storybook/utilities";

import FilterImage from ".";

const meta: ComponentMeta<typeof FilterImage> = {
  argTypes: {
    className: argTypes.className,
    image: {
      type: {
        name: "string",
        required: true,
      },
      control: "text",
      description: "URL of the image to be displayed",
      table: {
        type: {
          summary: "string",
        },
      },
    },
    brightness: {
      type: {
        name: "number",
      },
      control: { type: "number", min: 0, step: 0.1 },
      description:
        "Image brightness linear multiplier value. Applied to image via CSS `brightness()` function.",
      table: {
        type: {
          summary: "number",
        },
        category: "Filters",
        defaultValue: {
          summary: 1,
        },
      },
    },
    height: {
      type: {
        name: "number",
      },
      control: { type: "number", min: 0, step: 1 },
      description: "Height of the `canvas` element in `px`",
      table: {
        type: {
          summary: "number",
        },
        defaultValue: {
          summary: 600,
        },
      },
    },
    width: {
      type: {
        name: "number",
      },
      control: { type: "number", min: 0, step: 1 },
      description: "Width of the `canvas` element in `px`",
      table: {
        type: {
          summary: "number",
        },
        defaultValue: {
          summary: 600,
        },
      },
    },
    color: {
      type: {
        name: "string",
      },
      control: "color",
      description:
        "Color filter applied to the image by drawing over top of it with multiply composition.",
      table: {
        type: {
          summary: "string",
        },
        category: "Filters",
        defaultValue: {
          summary: "transparent",
        },
      },
    },
  },
  component: FilterImage,
};
export default meta;

export const Primary: ComponentStoryObj<typeof FilterImage> = {
  args: {
    height: 300,
    image:
      "https://rubin.canto.com/direct/image/3jhaht4beh6aha6eiplkb5us6o/MlO1oSfjiMJBT25-VtnRD7B0-EI/original?content-type=image%2Fjpeg&name=Wide_View_Telescope_Mount.jpg",
    color: "#EC1C24",
  },
};
