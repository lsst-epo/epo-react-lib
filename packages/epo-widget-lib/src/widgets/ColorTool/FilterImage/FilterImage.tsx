import "context-filter-polyfill";
import { FunctionComponent, HTMLAttributes, useRef } from "react";
import useImage from "@/hooks/useImage";
import { isStyleSupported } from "../utilities";
import * as Styled from "../styles";

export interface FilterImageProps extends HTMLAttributes<HTMLCanvasElement> {
  className?: string;
  height?: number;
  width?: number;
  url: string;
  color?: string;
  brightness?: number;
  active: boolean;
  filters?: {
    [key: string]: number | undefined;
  };
  onLoadCallback?: () => void;
}

const FilterImage: FunctionComponent<FilterImageProps> = ({
  width = 600,
  height = 600,
  className,
  url,
  color = "transparent",
  filters = {
    brightness: 1,
  },
  active,
  onLoadCallback,
}) => {
  const [image, status] = useImage({
    url,
    crossOrigin: "anonymous",
    onLoadCallback,
  });
  const canvasRef = useRef<HTMLCanvasElement>(null);
  const ctx = canvasRef.current?.getContext("2d");

  const updateColor = (
    ctx: CanvasRenderingContext2D,
    color: string,
    canvasWidth: number,
    canvasHeight: number
  ) => {
    const safeColor = isStyleSupported("color", color) ? color : "transparent";

    ctx.fillStyle = safeColor;
    ctx.fillRect(0, 0, canvasWidth, canvasHeight);
  };

  const getFilters = (filters: {
    [key: string]: number | undefined;
  }): string => {
    return Object.keys(filters).reduce((prev, key, i) => {
      return (prev += `${i > 0 ? " " : ""}${key}(${filters[key]})`);
    }, "");
  };

  if (ctx && image && status === "loaded") {
    const { width: canvasWidth, height: canvasHeight } = ctx.canvas;
    ctx.clearRect(0, 0, canvasWidth, canvasHeight);

    ctx.globalCompositeOperation = "multiply";
    ctx.filter = getFilters(filters);
    ctx.drawImage(image, 0, 0, canvasWidth, canvasHeight);

    updateColor(ctx, color, canvasWidth, canvasHeight);
  }

  return (
    <Styled.Image
      ref={canvasRef}
      role="img"
      style={{
        "--image-visibility": active ? "visible" : "hidden",
        "--image-opacity": active ? 1 : 0,
      }}
      {...{ className, width, height }}
    />
  );
};

FilterImage.displayName = "Widgets.ColorTool.FilterImage";

export default FilterImage;
