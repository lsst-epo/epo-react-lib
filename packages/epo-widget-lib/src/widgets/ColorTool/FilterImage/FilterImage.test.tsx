import { act, render, screen } from "@testing-library/react";
import FilterImage from ".";

const props = {
  image: "https://via.placeholder.com/300",
  width: 300,
  height: 300,
};

describe("FilterImage", () => {
  it(`should compile`, () => {
    render(<FilterImage {...props} />);

    const canvas = screen.getByRole("img");

    expect(canvas).toBeInTheDocument();
    expect(canvas).toHaveAttribute("width", props.width.toString());
    expect(canvas).toHaveAttribute("height", props.height.toString());
  });
  it("should render an image", async () => {
    await act(async () => {
      render(<FilterImage {...props} />);
    });

    const canvas = screen.getByRole<HTMLCanvasElement>("img");
    const img = canvas.toDataURL();

    expect(img).toMatchSnapshot();
  });
});
