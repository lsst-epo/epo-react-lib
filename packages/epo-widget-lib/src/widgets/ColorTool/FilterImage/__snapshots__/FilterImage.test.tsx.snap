// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterImage should render an image 1`] = `"data:image/png;base64,00"`;
